src/webgpu/api/validation/copyBetweenLinearDataAndTexture.spec.ts

```diff
@@ -0,0 +1,745 @@
+      return bytesPerImage * (copyExtent.depth - 1) + bytesInLastSlice;
+    }
+  }
+
+  testRun(options: {
+    texture: GPUTexture;
+    origin: GPUOrigin3D;
+    size: GPUExtent3D;
+    mipLevel?: number;
+    offset?: number;
+    bytesPerRow: number;
+    rowsPerImage: number;
+    dataSize: number;
+    method: TestMethod;
+  }): void {
+    const {
+      texture,
+      origin,
+      size,
+      mipLevel = 0,
+      offset = 0,
+      bytesPerRow,
+      rowsPerImage,
+      dataSize,
+      method,
+    } = options;
+
+    switch (method) {
+      case TestMethod.WriteTexture: {
+        const data = new Uint8Array(dataSize);
+
+        this.device.defaultQueue.writeTexture(
+          { texture: texture, mipLevel: mipLevel, origin: origin },
+          data,
+          { offset: offset, bytesPerRow: bytesPerRow, rowsPerImage: rowsPerImage },
+          size
+        );
+
+        break;
+      }
+      case TestMethod.CopyBufferToTexture: {
+        const buffer = this.device.createBuffer({
+          mappedAtCreation: false,
+          size: dataSize,
+          usage: GPUBufferUsage.COPY_SRC,
+        });
+
+        const encoder = this.device.createCommandEncoder();
+        encoder.copyBufferToTexture(
+          { buffer: buffer, offset: offset, bytesPerRow: bytesPerRow, rowsPerImage: rowsPerImage },
+          { texture: texture, mipLevel: mipLevel, origin: origin },
+          size
+        );
+        this.device.defaultQueue.submit([encoder.finish()]);
+
+        break;
+      }
+      case TestMethod.CopyTextureToBuffer: {
+        const buffer = this.device.createBuffer({
+          mappedAtCreation: false,
+          size: dataSize,
+          usage: GPUBufferUsage.COPY_DST,
+        });
+
+        const encoder = this.device.createCommandEncoder();
+        encoder.copyTextureToBuffer(
+          { texture: texture, mipLevel: mipLevel, origin: origin },
+          { buffer: buffer, offset: offset, bytesPerRow: bytesPerRow, rowsPerImage: rowsPerImage },
+          size
+        );
+        this.device.defaultQueue.submit([encoder.finish()]);
+
+        break;
+      }
+    }
+  }
+}
+
+export const g = makeTestGroup(CopyBetweenLinearDataAndTextureTest);
+
+/*** Texture-related validation: ***/
+
+g.test('texture_must_be_valid')
+  .params(
+    params()
+      .combine(poptions('method', [TestMethod.WriteTexture, TestMethod.CopyBufferToTexture, TestMethod.CopyTextureToBuffer]))
+      .combine(pbool('success'))
+  )
+  .fn(async t => {
+    const { method, success } = t.params;
+
+    const texture = t.device.createTexture({
+      size: { width: 4, height: 4, depth: 1 },
+      mipLevelCount: 1,
+      sampleCount: 1,
+      format: 'rgba8unorm',
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.COPY_DST,
+    });
+
+    if (!success) {
+      // The texture becomes invalid.
+      texture.destroy();
+    }
+
+    t.expectValidationError(() => {
```

Comment:
Just to avoid boilerplate, maybe consider folding the expectValidationError call into testRun().

Reply:
Done